src/rgw/rgw_rados.cc

```diff
@@ -12868,33 +12850,30 @@ int RGWRados::cls_obj_usage_log_trim(string& oid, string& user, uint64_t start_e
   return r;
 }
 
-int RGWRados::remove_objs_from_index(RGWBucketInfo& bucket_info, list<rgw_obj_index_key>& oid_list)
+int RGWRados::remove_objs_from_index(RGWBucketInfo& bucket_info, list<rgw_obj>& oid_list)
 {
-  librados::IoCtx index_ctx;
-  string dir_oid;
-
   uint8_t suggest_flag = (get_zone().log_data ? CEPH_RGW_DIR_SUGGEST_LOG_OP : 0);
 
-  int r = open_bucket_index(bucket_info, index_ctx, dir_oid);
-  if (r < 0)
-    return r;
-
-  bufferlist updates;
-
   for (auto iter = oid_list.begin(); iter != oid_list.end(); ++iter) {
+    bufferlist updates, out;
     rgw_bucket_dir_entry entry;
-    entry.key = *iter;
+    iter->key.get_index_key(&entry.key);
     dout(2) << "RGWRados::remove_objs_from_index bucket=" << bucket_info.bucket << " obj=" << entry.key.name << ":" << entry.key.instance << dendl;
     entry.ver.epoch = (uint64_t)-1; // ULLONG_MAX, needed to that objclass doesn't skip out request
     updates.append(CEPH_RGW_REMOVE | suggest_flag);
     ::encode(entry, updates);
+    RGWRados::BucketShard bs(this);
```

Comment:
now we are running dir_suggest one object at a time this won't scale when we have lots of objects.
remove_objs_from_index is called in two cases:
1. radosgw-admin obj unlink - that is used for a single object and than there is no scaling issue (and which fails)
2. check_bad_index_multipart (called when listing a bucket and when running bucket check --fix) here the scale is a major factor.

The only solution I see is using a different function for radosgw-admin obj unlink.
You will need to add a new method remove_obj_from_index that gets a single object and removes it (using the BucketShard).

Reply:
@oritwas  I think there is no need to add a new method, radosgw-admin obj unlink can use this method too.  In radosgw-admin obj unlink the oid_list size is always 1.

Reply:
@shashalu, the problem is the changes effect object list and bucket check commands. those needs the performance and scale as they call the methods with a long list of objects